### PR TITLE
Broadcast with dims matcher

### PR DESCRIPTION
**Pull request by bpickrel**
Add a matcher to `simplify_dyn_ops` which replaces `broadcast_with_dims` (Onnx `Expand`) with the equivalent `multibroadcast` with a fixed output shape attribute.  

This conversion is not just a simplification but is required for this op. to run at all since a definite output shape is needed at runtime.  `broadcast_with_dims` has a dynamic output shape at compile time, and this matcher replaces it with a fixed value that is available at runtime.

### CODE REVIEW

**Comment by CharlieL7** [comment]
> This conversion is not just a simplification but is required for this op. to run at all since a definite output shape is needed at runtime. `broadcast_with_dims` has a dynamic output shape at compile time, and this matcher replaces it with a fixed value that is available at runtime.

That's not a fully correct statement. `broadcast_with_dims` will run fine on the `ref` target. This matcher is required for dynamic batch and to run on GPU.

File: src/simplify_dyn_ops.cpp [hunk 1]
```diff
@@ -26,12 +26,141 @@
...
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
+/**
+ *    Convert a Broadcast_with_dims op, which always outputs a dynamic shape,
+ *    to a Multibroadcast op with a static output shape attribute.
+ *
+ *    This conversion is required to be made since runtime target ops can only output
+ *    static shapes.
```
Comment (CharlieL7) [inline src/simplify_dyn_ops.cpp:41]: Remove

File: src/simplify_dyn_ops.cpp [hunk 2]
```diff
@@ -26,12 +26,141 @@
...
+#include <migraphx/op/resize.hpp>
 #include <migraphx/common.hpp>
 #include <migraphx/tensor_view.hpp>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
+/**
+ *    Convert a Broadcast_with_dims op, which always outputs a dynamic shape,
+ *    to a Multibroadcast op with a static output shape attribute.
```
Comment (CharlieL7) [inline src/simplify_dyn_ops.cpp:38]:
```suggestion
 *    Replace broadcast_with_dims operators with a static input tensor and a constant `dims` input
 *    into multibroadcast op with a static output shape attribute.
```

**Review by CharlieL7 — CHANGES_REQUESTED** [review]
Need to update this PR to merge some changes

File: test/op_shape_test.cpp [hunk 3]
```diff
@@ -966,6 +966,30 @@ TEST_CASE(broadcast_for_dot_dyn2)
         s0);
 }
 
+TEST_CASE(broadcast_with_dims0)
```
Comment (bpickrel) [inline test/op_shape_test.cpp:969]: @CharlieL7 I think these two tests were from you, but Github is showing them as new to this PR.  Did you mean to keep them, or not?

**Review by lakhinderwalia — APPROVED** [review]
1. Please add a negative test, with dimensions that cannot be multi-broadcasted. 
2. Approving the PR.
Thanks.

**Comment by bpickrel** [comment]
@CharlieL7 , Chris asked that we merge this PR into #2799 instead of waiting for that one to be merged into `develop`.  It shouldn't cause much extra fuss, but I want to make sure you're OK before I do it.  Recall that any of us has permissions to make a merge that isn't into the develop branch.

File: src/simplify_dyn_ops.cpp [hunk 4]
```diff
@@ -33,6 +33,36 @@
...
+                                                  match::arg(1)(match::is_constant()));
+    }
+
+    void apply(module& m, const match::matcher_result& mr) const
+    {
+        auto ins    = mr.result;
+        auto inputs = ins->inputs();
+
+        // read the values of arg(1) to create input to multibroadcast
+        std::vector<size_t> sizes_vec(inputs.at(0)->get_shape().ndim());
```
Comment (pfultz2) [inline src/simplify_dyn_ops.cpp:56]: Dont need to set the size of the vector, it will already get set in the `assign` function below.

Comment (bpickrel) [reply src/simplify_dyn_ops.cpp]: changed

**Comment by CharlieL7** [comment]
> @CharlieL7 , Chris asked that we merge this PR into #2799 instead of waiting for that one to be merged into `develop`. It shouldn't cause much extra fuss, but I want to make sure you're OK before I do it. Recall that any of us has permissions to make a merge that isn't into the develop branch.

Yeah, that's fine

File: test/op_shape_test.cpp (hunk 3 above)
Comment (CharlieL7) [reply test/op_shape_test.cpp:945]: The tests look the same to me, just moved around.

File: test/simplify_dyn_ops_test.cpp [hunk 5]
```diff
@@ -34,6 +34,36 @@ void run_pass(migraphx::module& m)
     migraphx::run_passes(m, {migraphx::simplify_dyn_ops{}, migraphx::dead_code_elimination{}});
 }
 
+TEST_CASE(broadcast_with_dims)
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply test/simplify_dyn_ops_test.cpp:37]: Looks good as is

**Review by CharlieL7 — APPROVED** [review]

**Comment by causten** [comment]
@bpickrel  please fix the merge conflict

**Pull request merged**